File: ccc2007j5.py
#Solution by Srihari Vishnu
mindist = input()
maxdist = input()
N = input()
#0, 990, 1010, 1970, 2030, 2940, 3060, 3930, 4060, 4970, 5030, 5990, 6010, 7000
count = 0
def between(curr, y):
  if y - curr >= mindist and y - curr <= maxdist:
    return True
def findNext(current):
  global count
  if current == 7000:
    count += 1
  else:
    for x in range(current, len(motels)):
      if motels[x]:
        if between(current, x):
          findNext(x)
      
motels = [False for x in range(7001)]
motels[0] = True
motels[990] = True
motels[1010] = True
motels[1970] = True
motels[2030] = True
motels[2940] = True
motels[3060] = True
motels[3930] = True
motels[4060] = True
motels[4970] = True
motels[5030] = True
motels[5990] = True
motels[6010] = True
motels[7000] = True

for x in range(N):
  motels[input()] = True
  
findNext(0)
print count

File: ccc2007j5.cpp
//Solution by Srihari Vishnu
//Dynamic Programming
//C++ Solution to CCC 2007 J5 : Keep on Truckin'

#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;
int main() {
  vector<int> motels;
  int ways[7001];
  int mindist, maxdist, n;
  int x;
  motels.push_back(0);
  motels.push_back(990);
  motels.push_back(1010);
  motels.push_back(1970);
  motels.push_back(2030);
  motels.push_back(2940);
  motels.push_back(3060);
  motels.push_back(3930);
  motels.push_back(4060);
  motels.push_back(4970);
  motels.push_back(5030);
  motels.push_back(5990);
  motels.push_back(6010);
  motels.push_back(7000);

  fill(ways, ways+7001, 0);
  cin >> mindist >> maxdist >> n;
  for (int i = 0; i < n; i++) {
    cin >> x;
    motels.push_back(x);
  }
  n += 14;
  sort(motels.begin(), motels.end());
  ways[0] = 1;
  for (int i = 0; i<motels.size(); i++) {
    for (int j = i; j < n; j++) {
      if (motels[j] >= motels[i]+ mindist && motels[j] <= motels[i]+maxdist) {
        ways[j] += ways[i];
      }
    }
  }
  cout << ways[n-1] << endl;

}
